File: app/lib/poe_flip_finder/gateways/ggg_exchange_source_gateway.ex
defmodule PoeFlipFinder.Gateways.GggExchangeSourceGateway do
  @moduledoc """
  The imperative-shell implementation of `PoeFlipFinder.ExchangeSourceGateway`,
  calling GGG's public Currency Exchange change-stream
  (docs/DATA_SOURCES.md § Currency Exchange Data). This is the
  Anti-Corruption Layer boundary: everything above this module only ever
  sees `PoeFlipFinder.ExchangeMarketEntry`, never GGG's raw JSON shape.

  Reaching the current, still-open hour is signaled by an HTTP 404 whose
  body echoes the requested change id back with an empty markets array --
  verified by direct testing 2026-08-06, not a failure.
  """

  @behaviour PoeFlipFinder.ExchangeSourceGateway

  alias PoeFlipFinder.{ExchangeChangeStreamPage, ExchangeMarketEntry}

  @default_base_url "https://web.poecdn.com"

  @impl true
  def fetch_hour(change_id) do
    url = base_url() <> "/api/currency-exchange/#{change_id}"

    # retry: false -- matches the Java gateway's behavior (a failure
    # propagates immediately; no automatic retry). Req retries on 5xx and
    # transport errors by default, which would also fight Bypass's
    # expect_once assertions in tests.
    case Req.get(url, retry: false) do
      {:ok, %Req.Response{status: 200, body: body}} ->
        {:ok, normalize(decode(body), change_id, false)}

      {:ok, %Req.Response{status: 404, body: body}} ->
        {:ok, normalize(decode(body), change_id, true)}

      {:ok, %Req.Response{status: status}} ->
        {:error, {:unexpected_status, status}}

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp base_url do
    Application.get_env(:poe_flip_finder, __MODULE__, [])[:base_url] || @default_base_url
  end

  # Req auto-decodes a JSON response body by content-type already, but a
  # Bypass-served response in tests may hand back a raw string -- accept
  # both so tests don't need to fake response headers just to satisfy this.
  defp decode(body) when is_binary(body), do: Jason.decode!(body)
  defp decode(body) when is_map(body), do: body

  @doc "Exposed so contract tests can exercise parsing without a live HTTP call."
  @spec normalize(map(), integer(), boolean()) :: ExchangeChangeStreamPage.t()
  def normalize(raw, requested_change_id, at_tip) do
    snapshot_hour = DateTime.from_unix!(requested_change_id)

    entries =
      raw
      |> Map.fetch!("markets")
      |> Enum.map(&to_entry(&1, snapshot_hour))

    %ExchangeChangeStreamPage{
      entries: entries,
      next_change_id: Map.fetch!(raw, "next_change_id"),
      at_tip: at_tip
    }
  end

  defp to_entry(raw, snapshot_hour) do
    [a, b] = raw["market_pair"]

    %ExchangeMarketEntry{
      league_external_id: raw["league"],
      currency_a_external_id: a,
      currency_b_external_id: b,
      snapshot_hour: snapshot_hour,
      volume_traded_a: long_value(raw["volume_traded"], a),
      volume_traded_b: long_value(raw["volume_traded"], b),
      lowest_stock_a: long_value(raw["lowest_stock"], a),
      highest_stock_a: long_value(raw["highest_stock"], a),
      lowest_stock_b: long_value(raw["lowest_stock"], b),
      highest_stock_b: long_value(raw["highest_stock"], b),
      lowest_ratio_a: double_value(raw["lowest_ratio"], a),
      highest_ratio_a: double_value(raw["highest_ratio"], a),
      lowest_ratio_b: double_value(raw["lowest_ratio"], b),
      highest_ratio_b: double_value(raw["highest_ratio"], b)
    }
  end

  defp long_value(values_by_item_path, item_path), do: Map.get(values_by_item_path, item_path, 0)

  # GGG's JSON omits the decimal point for whole-number ratios, so Jason
  # decodes those as integers -- forced to float here so callers always see
  # the same type Java's `double` deserialization would have produced.
  defp double_value(values_by_item_path, item_path) do
    values_by_item_path |> Map.get(item_path, 0) |> to_float()
  end

  defp to_float(value) when is_integer(value), do: value * 1.0
  defp to_float(value) when is_float(value), do: value
end

File: backend/src/main/java/com/poeflipfinder/backend/framework/exchange/GggExchangeSourceGateway.java
package com.poeflipfinder.backend.framework.exchange;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.poeflipfinder.backend.gateway.ExchangeChangeStreamPage;
import com.poeflipfinder.backend.gateway.ExchangeMarketEntry;
import com.poeflipfinder.backend.gateway.ExchangeSourceGateway;
import java.io.IOException;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestClient;
import org.springframework.web.client.RestClientException;

/**
 * The Frameworks & Drivers implementation of ExchangeSourceGateway, calling
 * GGG's public Currency Exchange change-stream (docs/DATA_SOURCES.md §
 * Currency Exchange Data). This is the Anti-Corruption Layer boundary:
 * everything above this class only ever sees ExchangeMarketEntry, never this
 * class's Raw* shapes.
 *
 * <p>Reaching the current, still-open hour is signaled by an HTTP 404 whose
 * body echoes the requested change id back with an empty markets array --
 * verified by direct testing 2026-08-06, not a failure.
 */
public class GggExchangeSourceGateway implements ExchangeSourceGateway {

    private final RestClient restClient;
    private final ObjectMapper objectMapper;

    public GggExchangeSourceGateway(RestClient.Builder restClientBuilder) {
        this.restClient = restClientBuilder.baseUrl("https://web.poecdn.com").build();
        this.objectMapper = new ObjectMapper();
    }

    @Override
    public ExchangeChangeStreamPage fetchHour(long changeId) {
        try {
            String json = restClient
                    .get()
                    .uri("/api/currency-exchange/{changeId}", changeId)
                    .retrieve()
                    .body(String.class);
            return normalize(json, changeId, false);
        } catch (HttpClientErrorException.NotFound e) {
            return normalize(e.getResponseBodyAsString(), changeId, true);
        } catch (RestClientException e) {
            throw new ExchangeSourceGatewayException("Failed to fetch Currency Exchange hour " + changeId, e);
        }
    }

    /**
     * Public so contract and integration tests can exercise real parsing
     * against a saved fixture without a live HTTP call -- including tests
     * outside this package that need real-data-through-real-parsing, e.g.
     * DivinationCardOpportunityFinderIntegrationTest.
     */
    public ExchangeChangeStreamPage normalize(String json, long requestedChangeId, boolean atTip) {
        RawPage rawPage = deserialize(json);
        Instant snapshotHour = Instant.ofEpochSecond(requestedChangeId);
        List<ExchangeMarketEntry> entries =
                rawPage.markets().stream().map(raw -> toEntry(raw, snapshotHour)).toList();
        return new ExchangeChangeStreamPage(entries, rawPage.nextChangeId(), atTip);
    }

    private RawPage deserialize(String json) {
        try {
            return objectMapper.readValue(json, RawPage.class);
        } catch (IOException e) {
            throw new ExchangeSourceGatewayException("Failed to parse Currency Exchange response", e);
        }
    }

    private ExchangeMarketEntry toEntry(RawMarket raw, Instant snapshotHour) {
        String a = raw.marketPair().get(0);
        String b = raw.marketPair().get(1);
        return new ExchangeMarketEntry(
                raw.league(),
                a,
                b,
                snapshotHour,
                longValue(raw.volumeTraded(), a),
                longValue(raw.volumeTraded(), b),
                longValue(raw.lowestStock(), a),
                longValue(raw.highestStock(), a),
                longValue(raw.lowestStock(), b),
                longValue(raw.highestStock(), b),
                doubleValue(raw.lowestRatio(), a),
                doubleValue(raw.highestRatio(), a),
                doubleValue(raw.lowestRatio(), b),
                doubleValue(raw.highestRatio(), b));
    }

    private long longValue(Map<String, Long> valuesByItemPath, String itemPath) {
        return valuesByItemPath.getOrDefault(itemPath, 0L);
    }

    private double doubleValue(Map<String, Double> valuesByItemPath, String itemPath) {
        return valuesByItemPath.getOrDefault(itemPath, 0.0);
    }

    // ignoreUnknown so fields GGG adds later (e.g. market_id, which we
    // recompute from market_pair rather than store) never break parsing.
    @JsonIgnoreProperties(ignoreUnknown = true)
    private record RawPage(@JsonProperty("next_change_id") long nextChangeId, List<RawMarket> markets) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    private record RawMarket(
            String league,
            @JsonProperty("market_pair") List<String> marketPair,
            @JsonProperty("volume_traded") Map<String, Long> volumeTraded,
            @JsonProperty("lowest_stock") Map<String, Long> lowestStock,
            @JsonProperty("highest_stock") Map<String, Long> highestStock,
            @JsonProperty("lowest_ratio") Map<String, Double> lowestRatio,
            @JsonProperty("highest_ratio") Map<String, Double> highestRatio) {
    }
}
